README.md
# React View Cache
cache react views.

## Introduction
With react-view-cache, you can render your components within a special **Cache** component. When you render another 
component within the same cache, the previous component will not unmount immediately, but just hide by being set as 
`display: none`, with all state reserved.

It's great to cache pages in a SPA. By saving and restoring scroll in cache hooks, you will get an optimized app. When 
you browse back and forward, the previous page shows up immediately with the old scroll without any reloading. More, you 
will also make less effort to manage data and states between pages.

## Basic Usage
```
import createCache from "react-view-cache";
const Cache = createCache();

...

const view = <Component ... />; // your component
const viewId = ...; // unique cache id for this component

...

<Cache viewId={viewId} view={view}/>
```

## Options
You can pass in options as the args of `createCache`, or modify them on the created `Cache` instance.
```
const Cache = createCache(options)
// or
Cache.options.xxx = ...
```

* cacheTime - number, milliseconds, default 5 * 60 * 1000
* cacheLimit - number, default 5
* hooks
  * beforeSwitch(oldViewId, newViewId)
  * afterSwitch(oldViewId, newViewId)
  
## Cache
Cache will render component by current props, and the previous components will be hidden.

### props
* viewId - unique cache id for this component
* view - one of the following types
  * react node
  * func({isActive}): react node
* [cacheTime] - optional, cache time for this component
  
## License
MIT

CHANGELOG.md
### 1.1.1

- introduce `prop-types` package

### 1.1.0

- support add/remove hooks
- fix bugs of switch hooks

### 1.0.0

First Version

test/src/entry.js
import ReactDOM from "react-dom";
import React from "react";
import createCache from "react-view-cache";

const Cache = createCache({
  hooks: {
    beforeSwitch(oldId, newId) {
      console.log('before switch', oldId, newId);
    },

    afterSwitch(oldId, newId) {
      console.log('after switch', oldId, newId)
    }
  }
});

class Page extends React.Component {
  state = {
    text: ''
  };

  render() {
    const { text } = this.state;

    return <div>
      <form onSubmit={::this.onSubmit}>
        <input ref="input"/>
        <button type="submit">ok</button>
      </form>

      <Cache viewId={text} view={({ isActive }) => <View isActive={isActive} text={text}/>}/>
    </div>
  }

  onSubmit(e) {
    e.preventDefault();
    this.setState({ text: this.refs.input.value });
  }
}

class View extends React.Component {
  componentDidMount() {
    console.log('mount', this.props.text);
  }

  componentWillUnmount() {
    console.log('unmount', this.props.text);
  }

  render() {
    const { text, isActive } = this.props;

    console.log('render', text, isActive);

    return <div>
      <div>text: {text}</div>
      <div>isActive: {isActive ? 'true' : 'false'}</div>
    </div>
  }
}

ReactDOM.render(
  <Page/>,
  document.getElementById('root')
);

test/webpack.config.js
var path = require('path');

module.exports = {
  entry: "./src/entry.js",
  output: {
    path: __dirname,
    filename: "./dist/bundle.js"
  },
  module: {
    loaders: [
      { test: /\.js$/, exclude: /node_modules/, loader: "babel-loader" }
    ]
  },
  resolve: {
    root: [
      path.resolve(__dirname, '../../'),
    ]
  },
  resolveLoader: {
    root: [
      path.resolve(__dirname, './node_modules')
    ]
  },
  devtool: 'source-map'
};

src/lib.js
export function deepEqual(a, b) {
  if (!isObject(a) || !isObject(b)) return a === b;
  else return deepEqualInner(a, b) && deepEqualInner(b, a);
}

function deepEqualInner(a, b) {
  for (let key in a) {
    if (!deepEqual(a[ key ], b[ key ])) return false;
  }
  return true;
}

function isObject(x) {
  return typeof x === 'object' && x !== null;
}

export function deepAssign(base, extension) {
  if (!isObject(base)) base = {};
  if (!isObject(extension)) extension = {};

  for (let key in extension) {
    if (!isObject(extension[ key ])) base[ key ] = extension[ key ];
    else base[ key ] = deepAssign(base[ key ], extension[ key ]);
  }

  return base;
}

export function getIds(items) {
  return items.map(item => item.id);
}

export function difference(a, b) {
  const result = [];
  for (let value of a) {
    if (b.indexOf(value) < 0) result.push(value);
  }
  return result;
}

src/index.js
import React from "react";
import {deepEqual, deepAssign, getIds, difference} from "./lib";
import PropTypes from 'prop-types';

const DefaultOptions = {
  cacheTime: 5 * 60 * 1000, // 5 minutes
  cacheLimit: 5,
  hooks: {
    beforeSwitch(oldViewId, newViewId){
    },
    afterSwitch(oldViewId, newViewId){
    },
    beforeAdd(viewId){
    },
    afterAdd(viewId){
    },
    beforeRemove(viewId){
    },
    afterRemove(viewId){
    },
  }
};

export default function createCache(options) {
  class Cache extends React.Component {
    static options = deepAssign(deepAssign({}, DefaultOptions), options);

    static propTypes = {
      viewId: PropTypes.string.isRequired,
      view: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.node,
      ]).isRequired,
      cacheTime: PropTypes.number
    };

    state = {
      activeId: '',

      // {id, view, cacheTime, cacheTimer, lastActivatedAt}
      viewItems: [],
    };

    componentDidMount() {
      this.refresh(this.props);
    }

    componentWillReceiveProps(nextProps) {
      this.refresh(nextProps);
    }

    componentWillUnmount() {
      const { viewItems } = this.state;
      viewItems.forEach(item => clearTimeout(item.cacheTimer));

      Cache.options.hooks.beforeSwitch(this.state.activeId, '');
      Cache.options.hooks.afterSwitch(this.state.activeId, '');

      getIds(viewItems).forEach(id => {
        Cache.options.hooks.beforeRemove(id);
        Cache.options.hooks.afterRemove(id);
      })
    }

    componentWillUpdate(nextProps, nextState) {
      const oldId = this.state.activeId;
      const newId = nextState.activeId;
      const oldIds = getIds(this.state.viewItems);
      const newIds = getIds(nextState.viewItems);
      const addedIds = difference(newIds, oldIds);
      const removedIds = difference(oldIds, newIds);

      if (oldId !== newId) Cache.options.hooks.beforeSwitch(oldId, newId);
      addedIds.forEach(id => Cache.options.hooks.beforeAdd(id));
      removedIds.forEach(id => Cache.options.hooks.beforeRemove(id));
    }

    componentDidUpdate(prevProps, prevState) {
      const oldId = prevState.activeId;
      const newId = this.state.activeId;
      const oldIds = getIds(prevState.viewItems);
      const newIds = getIds(this.state.viewItems);
      const addedIds = difference(newIds, oldIds);
      const removedIds = difference(oldIds, newIds);

      if (oldId !== newId) Cache.options.hooks.afterSwitch(oldId, newId);
      addedIds.forEach(id => Cache.options.hooks.afterAdd(id));
      removedIds.forEach(id => Cache.options.hooks.afterRemove(id));
    }

    render() {
      const { activeId, viewItems } = this.state;

      return (
        <div>
          {
            viewItems.map(item => {
              const isActive = item.id === activeId;
              return <div key={item.id} style={{ display: isActive ? undefined : 'none' }}>
                {
                  typeof item.view === 'function' ?
                    <View args={{ isActive }} view={item.view}/> : item.view
                }
              </div>
            })
          }
        </div>
      )
    }

    refresh(props) {
      const { viewId, view, cacheTime } = props;
      const { activeId, viewItems } = this.state;

      const newViewItems = viewItems.slice();

      if (!viewId) {
        // set cache timer for switch-out item
        const preViewItem = newViewItems.find(item => item.id === activeId);
        this.setCacheTimer(preViewItem);
      }
      else {
        const viewItem = newViewItems.find(item => item.id === viewId);
        if (viewItem) {
          // update item
          viewItem.view = view;
          viewItem.cacheTime = cacheTime;
          viewItem.lastActivatedAt = new Date;

          if (viewId !== activeId) {
            // remove cache timer for switch-in item
            clearTimeout(viewItem.cacheTimer);

            // set cache timer for switch-out item
            const preViewItem = newViewItems.find(item => item.id === activeId);
            this.setCacheTimer(preViewItem)
          }
        }
        else {
          // create new item
          const newViewItem = {
            id: viewId,
            view: view,
            cacheTime: cacheTime,
            lastActivatedAt: new Date,
          };
          newViewItems.push(newViewItem);

          // set cache timer for previous item
          const preViewItem = newViewItems.find(item => item.id === activeId);
          this.setCacheTimer(preViewItem);

          // remove oldest surplus item
          if (newViewItems.length > this.getCacheLimit()) {
            const oldestItem = newViewItems.reduce(
              (min, cur) => cur.lastActivatedAt < min.lastActivatedAt ? cur : min,
              newViewItems[ 0 ]
            );
            if (oldestItem) {
              clearTimeout(oldestItem.cacheTimer);
              newViewItems.splice(newViewItems.findIndex(item => item === oldestItem), 1);
            }
          }
        }
      }

      // update state
      this.setState({
        activeId: viewId,
        viewItems: newViewItems,
      })
    }

    expireItem(id) {
      const { viewItems } = this.state;
      const viewItem = viewItems.find(item => item.id === id);
      if (viewItem) {
        clearTimeout(viewItem.cacheTimer);
        const newViewItems = viewItems.filter(item => item !== viewItem);
        this.setState({ viewItems: newViewItems });
      }
    }

    setCacheTimer(item) {
      if (item) {
        clearTimeout(item.cacheTimer);
        const cacheTime = this.getCacheTime(item);
        item.cacheTimer = setTimeout(() => this.expireItem(item.id), cacheTime);
      }
    }

    getCacheTime(item) {
      if (item.cacheTime !== undefined) return item.cacheTime;
      else return Cache.options.cacheTime;
    }

    getCacheLimit() {
      return Cache.options.cacheLimit;
    }
  }

  // optimize re-render
  class View extends React.Component {
    static propTypes = {
      args: PropTypes.object,
      view: PropTypes.func.isRequired,
    };

    shouldComponentUpdate(nextProps) {
      return !deepEqual(this.props, nextProps);
    }

    render() {
      const { args, view } = this.props;
      return view(args);
    }
  }

  return Cache;
}
